### PR TITLE
Add ChatModel, LLM, and Embeddings for Google's PaLM APIs

### PR DESCRIPTION
**Pull request by brotchie**
- Add langchain.llms.GooglePalm for text completion,
 - Add langchain.chat_models.ChatGooglePalm for chat completion,
 - Add langchain.embeddings.GooglePalmEmbeddings for sentence embeddings,
 - Add example field to HumanMessage and AIMessage so that users can feed in examples into the PaLM Chat API,
 - Add system and unit tests.

Note async completion for the Text API is not yet supported and will be included in a future PR.

Happy for feedback on any aspect of this PR, especially our choice of adding an example field to Human and AI Message objects to enable passing example messages to the API.

### CODE REVIEW

**Comment by brotchie** [comment]
Unit tests are failing because our genai Python package isn't a dependency. Will sort this out tomorrow.

**Comment by scafati98** [comment]
Big milestone!!

File: langchain/llms/google_palm.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+        values["client"] = text
+
+        if values["temperature"] is not None and not 0 <= values["temperature"] <= 1:
+            raise ValueError("temperature must be in the range [0.0, 1.0]")
+
+        if values["top_p"] is not None and not 0 <= values["top_p"] <= 1:
+            raise ValueError("top_p must be in the range [0.0, 1.0]")
+
+        if values["top_k"] is not None and values["top_k"] <= 0:
```
Comment (scafati98) [inline langchain/llms/google_palm.py:74]: Should we assert that this number is <= 40?

File: langchain/llms/google_palm.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            raise ValueError("top_k must be positive")
+
+        if values["max_output_tokens"] is not None and values["max_output_tokens"] <= 0:
```
Comment (scafati98) [inline langchain/llms/google_palm.py:77]: Should we verify that this num is <= 1000?

File: langchain/llms/google_palm.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            values, "google_api_key", "GOOGLE_API_KEY"
+        )
+        try:
+            import google.generativeai as genai
+            from google.generativeai import text
+
+            genai.configure(
+                api_key=google_api_key,
+                client_options={
+                    "api_endpoint": "autopush-generativelanguage.sandbox.googleapis.com"
```
Comment (dglazkov) [inline langchain/llms/google_palm.py:60]: Probably remove this line. Don't need it in production.

**Review by fyhao — APPROVED** [review]

File: langchain/chat_models/google_palm.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+    HumanMessage,
+    SystemMessage,
+)
+from langchain.utils import get_from_dict_or_env
+
+try:
+    # Import only for type checking.
+    import google.generativeai as genai
+except ImportError:
+    pass
```
Comment (dev2049) [inline langchain/chat_models/google_palm.py:22]: for type checking you can do something like
```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
  import google.generativeai as genie
```

File: langchain/embeddings/google_palm.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+    @root_validator()
+    def validate_environment(cls, values: Dict) -> Dict:
+        """Validate api key, python package exists."""
+        google_api_key = get_from_dict_or_env(
+            values, "google_api_key", "GOOGLE_API_KEY"
+        )
+        try:
+            import google.generativeai as genai
+            from google.generativeai import text
```
Comment (dev2049) [inline langchain/embeddings/google_palm.py:23]: stylistic: could we just use `genai.text`. 1 bc we already imported genai, 2 bit more readable (was confused by bare `text`)

File: langchain/schema.py [hunk 6]
```diff
@@ -79,6 +79,8 @@ def type(self) -> str:
 class HumanMessage(BaseMessage):
     """Type of message that is spoken by the human."""
 
+    example: bool = False
```
Comment (dev2049) [inline langchain/schema.py:82]: not sure notion of an example is generic enough to put on base abstraction. could we just pass in example message separately to `_messages_to_prompt_dict`?

File: langchain/chat_models/google_palm.py (hunk 4 above)
Comment (hwchase17) [reply langchain/chat_models/google_palm.py:22]: if you add `from __future__ import annotations` should work perfectly and remove need for quotes, etc

File: langchain/schema.py (hunk 6 above)
Comment (hwchase17) [reply langchain/schema.py:82]: i think fine to keep as is - if we want to move to base class later it should be easy enough

**Review by hwchase17 — COMMENTED** [review]
this looks great! minor comments on style stuff

larger comment: for discoverability i think it would be really really helpful to have an example notebook for all three models (similar to https://python.langchain.com/en/latest/modules/models/llms/integrations/cohere.html). practically speaking most of our users find different models and learn how to use them that way, so i would guess it would drastically improve adoption

File: langchain/schema.py (hunk 6 above)
Comment (dev2049) [reply langchain/schema.py:82]: i was saying maybe we don't need to put it on any class at all, but don't feel strongly

File: langchain/llms/google_palm.py (hunk 1 above)
Comment (brotchie) [reply langchain/llms/google_palm.py:68]: These are model specific, so I don't think we'd want to cap them prematurely?

File: langchain/llms/google_palm.py (hunk 2 above)
Comment (brotchie) [reply langchain/llms/google_palm.py:71]: Ditto re: model specificity.

File: langchain/llms/google_palm.py (hunk 3 above)
Comment (brotchie) [reply langchain/llms/google_palm.py]: Done

File: langchain/chat_models/google_palm.py (hunk 4 above)
Comment (brotchie) [reply langchain/chat_models/google_palm.py]: Adding `from __future__ import annotations` let me clean up some quotes and imports further down in the file. I still needed the TYPE_CHECKING to stop import failing if the user hasn't installed the genai Python package.

File: langchain/embeddings/google_palm.py (hunk 5 above)
Comment (brotchie) [reply langchain/embeddings/google_palm.py]: Turns out .generate_text and .generate_embeddings are already lifted onto the genai module anyway, so cleaned this up pretty well.

**Review by hwchase17 — COMMENTED** [review]
lgtm overall

again, notebooks would probably help for discorverability, but if you dont want to add just yet that's also fine - just lmk! happy to do whatever

**Comment by brotchie** [comment]
> lgtm overall
> 
> again, notebooks would probably help for discorverability, but if you dont want to add just yet that's also fine - just lmk! happy to do whatever

Let's merge this as-is. We'll send a follow up PR with a notebook examples for each of LLM, Chat Model, and Embeddings. (targeting Wednesday, 3rd May).

**Review by hwchase17 — APPROVED** [review]

**Pull request merged**